### PR TITLE
[CHI-365] CE Number input with step

### PR DESCRIPTION
**Pull request by janibolkvadze**
@jllr @dani-cl-madrid @SergioQCL @mattnickles 

https://ctl.atlassian.net/browse/CHI-365

### CODE REVIEW

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/272/ for more details.

**Comment by janibolkvadze** [comment]
test this please

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/273/ for more details.

**Comment by SergioQCL** [comment]
test this please

**Comment by allyourbotsarebelongtous** [comment]
Build FAILED. See http://jenkins.elasticbox.com/job/chi_pull_request/296/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/297/ for more details.

File: src/custom-elements/src/components/number-input/number-input.tsx [hunk 1]
```diff
@@ -77,39 +84,48 @@ export class NumberInput {
...
 
-      if (stepDifference !== 0) {
-        this.value = (
-          Math.round(((+this.value - stepDifference) * 10000) / 10000) +
-          this.step
-        ).toString();
-      }
+    if (newValue !== null) {
+        let steppedValue = Math.round(10000*((+newValue-this.initialValue)/this.step))/10000;
+        const steppedMax =  Math.round(10000*Math.floor((+this.max-this.initialValue)/this.step))/10000;
```
Comment (SergioQCL) [inline src/custom-elements/src/components/number-input/number-input.tsx]:
```suggestion
        const steppedMax = Math.round(10000*Math.floor((+this.max-this.initialValue)/this.step))/10000;
```

File: src/custom-elements/src/components/number-input/number-input.tsx [hunk 2]
```diff
@@ -77,39 +84,48 @@ export class NumberInput {
...
+        const steppedMin =  Math.round(10000*Math.ceil((+this.min-this.initialValue)/this.step))/10000;
```
Comment (SergioQCL) [inline src/custom-elements/src/components/number-input/number-input.tsx]:
```suggestion
        const steppedMin = Math.round(10000*Math.ceil((+this.min-this.initialValue)/this.step))/10000;
```

File: src/custom-elements/src/components/number-input/number-input.tsx [hunk 3]
```diff
@@ -77,39 +84,48 @@ export class NumberInput {
     CallbackQueue.queueProcess(this._didUpdateCallBackOnceQueue);
   }
 
-  handleChange(ev: Event) {
-    let stepDifference = 0;
+  static isNumeric (n: string | number) {
+      return ( typeof n === "number" || typeof n === "string" ) &&
```
Comment (SergioQCL) [inline src/custom-elements/src/components/number-input/number-input.tsx]:
```suggestion
      return (typeof n === "number" || typeof n === "string") &&
```

File: src/custom-elements/src/components/number-input/number-input.tsx [hunk 4]
```diff
@@ -77,39 +84,48 @@ export class NumberInput {
...
+          !isNaN( +n - parseFloat( `${n}` ) );
```
Comment (SergioQCL) [inline src/custom-elements/src/components/number-input/number-input.tsx]:
```suggestion
          !isNaN(+n - parseFloat(`${n}`));
```

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/304/ for more details.

File: src/custom-elements/src/components/number-input/number-input.tsx [hunk 5]
```diff
@@ -77,39 +84,48 @@ export class NumberInput {
     CallbackQueue.queueProcess(this._didUpdateCallBackOnceQueue);
   }
 
-  handleChange(ev: Event) {
-    let stepDifference = 0;
+  static isNumeric (n: string | number) {
+      return (typeof n === "number" || typeof n === "string") &&
```
Comment (dani-cl-madrid) [inline src/custom-elements/src/components/number-input/number-input.tsx]: Please, review ts-lint rules.

File: src/custom-elements/src/components/number-input/number-input.tsx [hunk 6]
```diff
@@ -77,39 +84,48 @@ export class NumberInput {
...
+    newValue = NumberInput.isNumeric(newValue) ? newValue : null;
 
-      if (stepDifference !== 0) {
-        this.value = (
-          Math.round(((+this.value - stepDifference) * 10000) / 10000) +
-          this.step
-        ).toString();
-      }
+    if (newValue !== null) {
+        let steppedValue = Math.round(10000*((+newValue-this.initialValue)/this.step))/10000;
```
Comment (dani-cl-madrid) [inline src/custom-elements/src/components/number-input/number-input.tsx]: Review TS Lint rules.

File: src/custom-elements/src/components/number-input/number-input.tsx [hunk 7]
```diff
@@ -77,39 +84,48 @@ export class NumberInput {
...
+        newValue = ((Math.round(steppedValue) * this.step) + (+this.initialValue)).toString();
     }
 
-    this._didUpdateCallBackOnceQueue.push(() => {
-      this.chiChange.emit(this.value);
-    });
+    if (newValue !== this.value) {
+        this._setNewValue(newValue);
+    } else if ((ev.target as HTMLInputElement).value.toString() !== this.value) {
+        this.forceRepaint = !this.forceRepaint;
```
Comment (dani-cl-madrid) [inline src/custom-elements/src/components/number-input/number-input.tsx]: What problems did you find that made you use this hack?

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/313/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/315/ for more details.

**Pull request merged**